In [1]:
import scipy
from scipy.optimize import linprog

In [2]:
scipy.__version__

'1.10.1'

# Programação linear inteira (PLI)

Um Problema de programação inteira é um modelo de programação linear no qual algumas ou todas as variáveis do problema pertencem ao conjunto dos números inteiros.
Quando todas as variáveis são inteiras, o modelo é denominado programação inteira pura.
Caso contrário, ele é denominado programação inteira mista (veja [MILP](https://en.wikipedia.org/wiki/Integer_programming) no Wikipedia).

A solução de um PLI é muito mais complexa que a programação linear onde os valores das variáveis de decisão são números reais, pois (intuitivamente) o fato de ter que ser um inteiro coloca muito mais restrição nas soluções possíveis. Este é um dos famosos [problema NP-difíceis](https://pt.wikipedia.org/wiki/NP-dif%C3%ADcil).

Para ver como resolver este problema com **linprog()**, vamos revisitar o problema que apresentamos no notebook [01-Alocacao de recursos.ipynb](https://github.com/h3dema/linear-programming/blob/main/01-Alocacao%20de%20recursos.ipynb).



### Recordando

O problema possui somente duas variaveis e o sistema de equações lineares é:


\begin{matrix}
maximize~z = & x + 2y \\
sujeito~a & 2x + y \leq 20 \\
 & -4x + 5y \leq 10 \\
 & -x + 2y \ge -2 \\
 & -x + 5y = 15 \\
 & x \ge 0 \\
 & y \ge 0
\end{matrix}


In [3]:
c = [-1, -2]
#      ─┬  ─┬
#       │   └┤ coeficiente de y
#       └────┤ coeficiente de  x

# Para as inequações:
A_ub = [[ 2,  1],  # coeficientes da restrição Eq.2 (lado esquerdo)
        [-4,  5],  # coeficientes da restrição Eq.2 (lado esquerdo)
        [ 1, -2]]  # coeficientes da restrição Eq.2 (lado esquerdo)

b_ub = [20,  # constante da restrição Eq.2 (lado direito)
        10,  # constante da restrição Eq.3 (lado direito)
        2]   # constante da restrição Eq.4 (lado direito)

# Para a equação (só Eq. 5)
A_eq = [[-1, 5]]  # Green constraint left side
b_eq = [15]       # Green constraint right side

bounds = [(0, float("inf")),  # Bounds of x
          (0, float("inf"))]  # Bounds of y

In [4]:
opt = linprog(c=c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq, bounds=bounds, method="highs")
opt.x

array([7.72727273, 4.54545455])

**Note**: Os valores retornados são números reais.

Se desejamos que seja estes valores sejam inteiros, por exemplo, porque este problema refere-se a produção de unidades de produtos e não vendemos um produto parcialmente construido, basta informar ao **linprog()** utilizando o parâmetro __integrality__. Contudo está opção só está disponíveis em versões novas (1.10+).

In [5]:
opt = linprog(c=c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq, bounds=bounds, method="highs", integrality=1)
opt.x

array([5., 4.])

Podemos ainda indicar explicitamente qual variavel é inteira

In [6]:
intvars = [1, 0]  # só a variável "x" precisa ser inteira

opt = linprog(c=c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq, bounds=bounds, method="highs", integrality=intvars)
opt.x

array([7. , 4.4])

In [7]:
intvars = [0, 1]  # só a variável "y" precisa ser inteira

opt = linprog(c=c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq, bounds=bounds, method="highs", integrality=intvars)
opt.x

array([5., 4.])